tempto-core/src/main/java/com/teradata/tempto/s3/S3Client.java

```diff
@@ -0,0 +1,384 @@
+        }
+
+        // If folder already exists, throw exception
+        if (exists(path)) {
+            throw new RuntimeException("Folder already exists - could not create folder: " + path);
+        }
+
+        // Create meta-data for your folder and set content-length to 0
+        ObjectMetadata metadata = new ObjectMetadata();
+        metadata.setContentLength(0);
+
+        // Create empty content
+        InputStream emptyContent = new ByteArrayInputStream(new byte[0]);
+
+        // Create a PutObjectRequest passing the folder name suffixed by /
+        PutObjectRequest putObjectRequest = new PutObjectRequest(bucketName,
+                path, emptyContent, metadata);
+
+        // Send request to S3 to create folder
+        client.putObject(putObjectRequest);
+
+        logger.debug("Created S3 folder {}", path);
+    }
+
+    /**
+     * Method to delete folder in S3.  This method calls the main deleteFolder method, setting the recursive
+     * flag to false
+     *
+     * @param path  folder path to delete
+     */
+    public void deleteFolder(String path)
+    {
+        deleteFolder(path, false);
+    }
+
+    /**
+     * Method to delete folder.  If recursive flag is set to true, all objects underneath this folder
+     * will also be deleted
+     *
+     * @param path      folder path to delete
+     * @param recursive flag if set to true will delete all objects underneath
+     */
+    public void deleteFolder(String path, boolean recursive)
+    {
+        // If no forward slash at end of path, add one
+        if (!path.substring(path.length() - 1).equals("/")) {
+            path += "/";
+        }
+
+        // Get all objects in this folder
+        ArrayList<String> objects = getAllObjects(path);
+
+        if (!recursive && objects.size() > 0) {
+            throw new RuntimeException("Folder is not empty - could not remove folder: " + path);
+        } else if (objects.size() > 0) {
+            for(String objectName : objects) {
+                deleteObject(objectName);
+            }
+        }
+
+        // Delete folder object
+        deleteObject(path);
+
+        logger.debug("Deleted S3 folder {}", path);
+    }
+
+    /**
+     * Method to save file from direct local file
+     *
+     * @param path      path on S3 to save
+     * @param filePath  path of input file to save
+     */
+    public void saveFile(String path, String filePath)
+    {
+        client.putObject(new PutObjectRequest(bucketName, path, new File(filePath)));
+
+        logger.debug("Saved file {} to S3", path);
+    }
+
+    /**
+     * Method to save file from input stream
+     *
+     * @param path          path on S3 to save
+     * @param stream        input stream
+     * @param byteLength    length of data (in bytes)
+     */
+    public void saveFile(String path, InputStream stream, long byteLength)
+    {
+        ObjectMetadata metadata = new ObjectMetadata();
+        metadata.setContentLength(byteLength);
```

Comment:
Do we need to know size of content upfront?
How do we validate that `stream` actually provider proper amount of data. Does wrapped `S3` client does that?